BDArmory/Ammo/PooledBullet.cs

```diff
@@ -506,6 +507,7 @@ void FixedUpdate()
         /// <param name="period">Period to consider, typically Time.fixedDeltaTime</param>
         public void MoveBullet(float period)
         {
+            atmosphereDensity = (float)FlightGlobals.getAtmDensity(FlightGlobals.getStaticPressure(currentPosition), FlightGlobals.getExternalTemperature(currentPosition)); //moving this here so it's not getting called 2x a tick in LeapFrogVelocityHalfStep
```

Comment:
GetDragAdjustedVelocity makes more sense, if it's underwater it's already getting overridden to 1030, so no need to calc density to begin with.

Reply:
`dragVelocityFactor` is very much a hack anyway, since it's based on the `atmDensity` at the point where it's asked about instead of being accumulated over the path of the bullet (`ρ` in the derivation comment is assumed to be constant, which it totally isn't for long distances or when switching between air and water).

Speaking of which, the
```csharp
currentVelocity *= dragVelocityFactor;
```
when the bullet is underwater isn't right as `dragVelocityFactor` is meant to account for the total change in velocity due to the drag over the entire path of the bullet up to that point, not the instantaneous deceleration due to drag, which ought to be something like `a = F / m = 1/2 * ρ * v^2 * Cd * A / m` (most of which could be pre-calculated as a constant).

Reply:
Wait, that's not quite right either.
`dragVelocityFactor` is the fraction to change the velocity by for the total time since `currentSpeed` was last updated, which happens every frame when it's underwater and on impacts, so `currentVelocity *= dragVelocityFactor;` being called every frame is fine when it's underwater since `currentSpeed` is also being updated every frame.
`currentSpeed` is really `initialSpeed` for each segment of calculating drag whenever `timeElapsedSinceCurrentSpeedWasAdjusted` is reset to 0 (variable naming could be better here...).

`UpdateDragEstimate` in `LeapfrogVelocityHalfStep` could be moved into the `if (underwater)` block as it's not used elsewhere until a hit, entering water or beehive detonation occurs, in which case `GetDragAdjustedVelocity` gets called (which needs to call `UpdateDragEstimate` to update `dragVelocityFactor`). Then the `atmDensity` calculation could be moved back into `UpdateDragEstimate` as it'd only be getting called above water on impacts/beehive detonations.

Reply:
Btw, the `GetDragAdjustedVelocity().magnitude + sBullet.bulletVelocity;` in `BeehiveDetonation` ought to be calculated only once outside the `for` loop, since that's the same for each subprojectile.